src/modules/fw_pos_control/performance_model/PerformanceModel.cpp

```diff
@@ -0,0 +1,225 @@
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+
+/**
+ * @file PerformanceModel.cpp
+ * Performance model.
+ */
+
+#include <geo/geo.h>
+#include <px4_platform_common/events.h>
+#include "PerformanceModel.h"
+
+// air density of standard athmosphere at 5000m above mean sea level [kg/m^3]
+static constexpr float kAirDensityStandardAtmos5000Amsl = 0.7363f;
+
+// air density of standard athmosphere at 1000m above mean sea level [kg/m^3]
+static constexpr float kAirDensityStandardAtmos1000Amsl = 1.112f;
+
+// [.] minimum ratio between the actual vehicle weight and the vehicle nominal weight (weight at which the performance limits are derived)
+static constexpr float kMinWeightRatio = 0.5f;
+
+// [.] maximum ratio between the actual vehicle weight and the vehicle nominal weight (weight at which the performance limits are derived)
+static constexpr float kMaxWeightRatio = 2.0f;
+
+// climbrate defining the service ceiling, used to compensate max climbrate based on air density
+static constexpr float kClimbrateMin = 0.5f; // [m/s]
+
+PerformanceModel::PerformanceModel(): ModuleParams(nullptr)
+{
+	updateParams();
+}
+float PerformanceModel::getWeightRatio() const
+{
+	float weight_factor = 1.0f;
+
+	if (_param_weight_base.get() > FLT_EPSILON && _param_weight_gross.get() > FLT_EPSILON) {
+		weight_factor = math::constrain(_param_weight_gross.get() / _param_weight_base.get(), kMinWeightRatio,
+						kMaxWeightRatio);
+	}
+
+	return weight_factor;
+}
+float PerformanceModel::getMaximumClimbRate(float air_density) const
+{
+
+	float climbrate_max = _param_fw_t_clmb_max.get();
+
+	const float density_min = _param_density_min.get();
+
+	if (density_min < kAirDensityStandardAtmos1000Amsl
```

Comment:
Why are these checks required? Can you add some comments? Is the 5000 limit required?

Reply:
@sfuhrer It's a protection against unfeasible value. I agree that the limits might not be extremely well chosen. I think the 5000m limit is actually more problematic since there are probably fixed wings that have their service ceiling higher than that (not sure about it though). But we also have to acknowledge that most of our current code assumes that the propulsion consists of a propeller driven by an electric motor in voltage control mode (RPM mode).
What I could do instead of disabling the entire compensation is to clamp the value.

Reply:
In other words, the protection requires the service ceiling of the vehicle to be between 1000 and 5000m msl.

Reply:
>What I could do instead of disabling the entire compensation is to clamp the value

Sounds better to me yeah.
I would extend the limit to 6000 or even more, I don't see why it needs to be constrained so much. Also the lower limit, why 1000 and not something smaller?